### DISCUSSION TITLE
How to debug Next.js core?

### DISCUSSION THREAD

**Discussion by jensmeindertsma** [vercel/next.js#13272] [Help] [answered]
I'm working on a pull request that changes the Next.js core. I'd like to be able to inspect the value of different variables at runtime. A `console.log` inside for example `next.js\packages\next\build\webpack\config\blocks\css\loaders\client.ts` does not work, and I'd like to use the VSCode debugger and set breakpoints. 

I've noticed that the root of the project does contain a `.vscode/launch.json` file (this suggests debugging is possible). Running `yarn dev` does not work, I suppose because it builds the code and doesn't execute it?

I've read the contributing guide and the debugging guide ( that is meant for bugs inside my Next.js app, not the core? ), but the contributing guide doesn't say anything about debugging the core.

Any help with this would be appreciated. Thanks in advance!

**Comment by jamesmosier** [comment, marked as answer]
Have you tried `yarn debug` from within the root directory? 

https://github.com/zeit/next.js/blob/9fb2432b14ed85889b892e5ba024fb07a204faf7/package.json#L36

I haven't tried this myself, but seems like a good starting point.

**Reply by jensmeindertsma** [reply]
Yeah I tried this but the build instantly fails because there is no `pages` directory. I tried running the `Launch app development` debugger to run an app ( so that there are pages), but it only triggers a few of my breakpoints. Can you help me with solving the `no pages` issue?

**Reply by timneutkens** [reply]
<img width="458" alt="Screen Shot 2020-05-24 at 17 34 50" src="https://user-images.githubusercontent.com/6324199/82758109-e2b77680-9de4-11ea-80cf-181ffc89b3ec.png">

If you use vscode you can boot up one of the integration test apps using the debugger inside of vscode.
In other cases you can run `yarn debug test/integration/basic` (or any other directory with a Next.js app in the monorepo, example: `yarn debug test/integration/production` (boots up in development mode) or `yarn debug build test/integration/production` `debug` === `next` (the Next.js cli)

**Reply by jensmeindertsma** [reply]
Thanks so much! I got it all working after removing the `.next` directory from `test/integration/basic` that had apparently cached the Babel plugins because the breakpoints inside the Babel plugin files were not triggering during compilation. It's a bit random when you don't remove the `.next` folder. Sometimes they get triggered, other times they don't.
